### DISCUSSION TITLE
Microsoft.PowerFx.Connectors

### DISCUSSION THREAD

**Discussion by rajyraman** [microsoft/Power-Fx#537] [General]
I had a look at these tests (https://github.com/microsoft/Power-Fx/blob/main/src/tests/Microsoft.PowerFx.Connectors.Tests/PowerPlatformConnectorTests.cs), but I am not able to connect the dots.

What is the intention of Microsoft.PowerFx.Connectors project?

**Comment by MikeStall** [comment]
Microsoft.PowerFx.Connectors  is support to enable Power Fx to call REST services, like what Power Apps does for Connector calls.  

It adds an extension method to config which adds services:
`config.AddService("FunctionNamespace", swagger, httpClient);`

swagger is an OpenAPI document that describes the rest service. These rest endpoints are then imported as functions into the given namespace. 
The swagger parsing is handled via https://github.com/Microsoft/OpenAPI.NET, which supports both swagger 2.0 and OpenAPI 3.0. 

Invoking the functions then invokes via the provided httpClient, which can handle authentication or any transforms between what the swagger describes and what the service actually expects. 

See these tests for calling an arbitrary REST service:
https://github.com/microsoft/Power-Fx/blob/main/src/tests/Microsoft.PowerFx.Connectors.Tests/BasicRestTests.cs

For example, 

For example, calling:
`config.AddService("Test", apiDocument, httpClient);`

Will inject functions, so that this Fx expression: 
`Test.GetWeatherWithHeader({ id : 11 })`

executes as a http call (via the provided httpClient) like this:
`GET http://localhost:5000/weather/header\r\n id: 11`



These dll is still in preview (we may make breaking changes), but you can certainly try it and give us feedback. 

This can also be used to call Power Platform connectors (https://docs.microsoft.com/en-us/connectors/connector-reference/connector-reference-powerapps-connectors), just like how Power Apps does. The test above demonstrates this, *however*, the challenge there is getting authentication permission to invoke these connectors. There's not a good way for 3rd parties to do this _yet_. 
We are exploring how to call connectors that are already created in the dataverse environment. Stay tuned on that one...

**Comment by rajyraman** [comment]
@MikeStall - That sounds great! I will look at the tests again and see how OAuth is handled.

**Reply by MikeStall** [reply]
The unit tests don't solve oauth. They just mock out a fake token. The host (ie, PowerApps) needs to provide the oauth token.
Are you trying to call the PowerApps Connectors or  just your own rest APIs?

**Reply by rajyraman** [reply]
@MikeStall - I plan to just test this with just plain Open API.
